Add long-running stability tests

dmitryax opened the pull request:
**Description:** 
End-2-end tests running for an hour with fixed resource limits. If limits are reached tests fail. 

Current state of the collector shows a slow grow in memory usage, which may indicate a memory leak that needs to be investigated. Once this solved, this tests can be upgraded to not use a fixed limit, but check resources consumption at the beginning and making sure it doesn't grow more than in a specified small percentage after that.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/1022

dmitryax commented on testbed/stabilitytests/.gitignore:
```diff
@@ -0,0 +1,2 @@
+results/*
+!results/.gitkeep
```
This file will not be needed once this PR is merged https://github.com/open-telemetry/opentelemetry-collector/pull/1116

tigrannajaryan commented on testbed/stabilitytests/metric_test.go:
```diff
@@ -0,0 +1,67 @@
+	scenarios.Scenario10kItemsPerSecond(
+		t,
+		testbed.NewOCMetricDataSender(testbed.GetAvailablePort(t)),
+		testbed.NewOCDataReceiver(testbed.GetAvailablePort(t)),
+		testbed.ResourceSpec{
+			ExpectedMaxCPU:      45,
+			ExpectedMaxRAM:      86,
+			ResourceCheckPeriod: resourceCheckPeriod,
+		},
+		nil,
+	)
+}
+
+func TestStabilityMetricsCarbon(t *testing.T) {
+	scenarios.Scenario10kItemsPerSecond(
+		t,
+		datasenders.NewCarbonDataSender(testbed.GetAvailablePort(t)),
+		datareceivers.NewCarbonDataReceiver(testbed.GetAvailablePort(t)),
+		testbed.ResourceSpec{
+			ExpectedMaxCPU:      237,
+			ExpectedMaxRAM:      90,
+			ResourceCheckPeriod: resourceCheckPeriod,
+		},
+		nil,
+	)
+}
+
+func TestStabilityMetricsSignalFx(t *testing.T) {
+	scenarios.Scenario10kItemsPerSecond(
+		t,
+		datasenders.NewSFxMetricDataSender(testbed.GetAvailablePort(t)),
+		datareceivers.NewSFxMetricsDataReceiver(testbed.GetAvailablePort(t)),
+		testbed.ResourceSpec{
+			ExpectedMaxCPU:      83,
+			ExpectedMaxRAM:      91,
+			ResourceCheckPeriod: resourceCheckPeriod,
+		},
+		nil,
+	)
+}
```
Do these need to be 3 separate functions or can be a matrix test like TestTrace10kSPS?

tigrannajaryan replied:
Can we also have OTLP test?

dmitryax replied:
Yes, we need them as separate tests. It makes it much easier to split between circleCI runners

tigrannajaryan commented on testbed/stabilitytests/metric_test.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+func TestStabilityMetricsOpenCensus(t *testing.T) {
+	scenarios.Scenario10kItemsPerSecond(
```
I was envisioning stability test different from Scenario10kItemsPerSecond.
Instead of checking for maximum CPU and RAM limits it would initially run for a few seconds and then would take the average peak value of CPU and peak value of RAM observed and then continue running for longer and see if CPU or RAM will go over the initially observed peak by more than a configured small amount. (Or perhaps we do this only for RAM and for CPU we keep the original monitoring logic).
What you have is fine for initial version but I would want to see a more sophisticated test in the future.

dmitryax replied:
Yes, that's much better approach. I was watching  the resource usage during the test. For most of the cases with current setup RAM usage doesn't seem to stabilize at one value. We probably need to introduce a memory limiter and make sure no drops happened, right?
You can see how it went btw https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/1279/workflows/4099caea-84f4-4cdd-9d75-2d1e4ed31165/jobs/3528

tigrannajaryan replied:
I am not sure why memory limiter is going to make a difference. Is it because it forces a GC when the limit is hit?

dmitryax replied:
Yes, I was thinking it may help. I've checked, and it doesn't matter. I'll need to investigate why do we have slowing growing RAM usage. Then I'll change the tests to work the way as you suggested

tigrannajaryan commented on testbed/datareceivers/carbon.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+// CarbonDataReceiver implements Carbon format receiver.
+type CarbonDataReceiver struct {
```
These are great, but perhaps introduce these new data receivers and senders in a separate PR to avoid mixing introduction of the stability test with them?

dmitryax replied:
Makes sense. Please take a look https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/320 
I'll rebase after

dmitryax committed:
Add long-running stability tests 

End-2-end tests running for an hour with fixed resource limits. If limits are reached tests fail.

Current state of the collector shows a slow grow in the memory usage, which may indicate a memory leak that needs to be investigated. Once this solved, this tests can be upgraded to not use a fixed limit, but check resources consumption at the beginning and making sure it doesn't grow more than in a specified small percentage after that.

tigrannajaryan commented:
@dmitryax I cannot find the results in CircleCi. Which job should show stability tests?

dmitryax commented:
The stability tests are triggered on master branch after merging, not per PR, so they are not visible here. I did some testing, here is an example https://app.circleci.com/jobs/github/open-telemetry/opentelemetry-collector-contrib/4036

tigrannajaryan reviewed (approved):
LGTM
